### PR TITLE
Feat/punctuation capitalization/long queries signoff

### PR DESCRIPTION
**Pull request by PeganovAnton**
Script `punctuate_capitalize.py` for inference on long queries.

### CODE REVIEW

File: tutorials/nlp/Punctuation_and_Capitalization.ipynb [hunk 1]
```diff
@@ -666,15 +641,148 @@
...
+   "cell_type": "code",
    "execution_count": null,
-   "outputs": []
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "inference_results = model.add_punctuation_capitalization(\n",
+    "    [\n",
+    "        \"each of these songs represents a scene a movement in some cases a sonic revolution that completely altered the \"\n",
+    "        \"course of popular music they're all also calling cards almost for those cities songs totally linked with their \"\n",
```
Comment (ekmb) [inline tutorials/nlp/Punctuation_and_Capitalization.ipynb:698]: What is the source of this text?

File: examples/nlp/token_classification/punctuate_capitalize.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        "predictions computation, margins are removed. In the next list, subtokens which logits are not used for final "
+        "predictions computation are marked with asterisk: ``[['[CLS]'*, 'h', 'e', 'l'*, '[SEP]'*], ['[CLS]'*, 'e'*, "
+        "'l', 'l'*, '[SEP]'*], ['[CLS]'*, 'l'*, 'l', 'o', '[SEP]'*]]``.",
+    )
+    parser.add_argument(
+        "--step",
+        "-s",
+        type=int,
+        default=8,
+        help="Relative shift of consequent segments into which long queries are split. Long queries are split into "
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize.py:128]: Are the defaults set here recommended values?

File: nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py [hunk 3]
```diff
@@ -423,44 +425,239 @@ def __getitem__(self, idx):
...
+    for j, word in enumerate(words):
+        word_tokens = tokenizer.text_to_tokens(word)
+        subtokens.extend(word_tokens)
+        subtokens_mask.append(1)
+        subtokens_mask.extend([0] * (len(word_tokens) - 1))
+    return subtokens, subtokens_mask
+
+
+def check_max_seq_length_and_margin_and_step(max_seq_length, margin, step):
+    if margin >= (max_seq_length - 2) // 2 and margin > 0 or margin < 0:
```
Comment (ekmb) [inline nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py:456]: same

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py [hunk 4]
```diff
@@ -365,92 +368,163 @@ def _setup_infer_dataloader(
...
+        ):
+            if not first:
+                new_start_word_ids[i] += torch.count_nonzero(stm[: margin + 1]).numpy()  # + 1 is for [CLS] token
+            stm = self.remove_margins(stm, margin, keep_left=first, keep_right=last)
+            for b_probs, logits in [(b_punct_probs, pl), (b_capit_probs, cl)]:
+                p = torch.nn.functional.softmax(
+                    self.remove_margins(logits, margin, keep_left=first, keep_right=last)[stm], dim=-1,
+                )
+                b_probs.append(p.detach().cpu().numpy())
+        return b_punct_probs, b_capit_probs, new_start_word_ids
```
Comment (ekmb) [inline nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py:437]: Please add docstrings and types to the newly added methods. How about using leading underscore in the names of all these functions?

**Review by ekmb — CHANGES_REQUESTED** [review]

File: nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py [hunk 5]
```diff
@@ -423,44 +425,239 @@ def __getitem__(self, idx):
         )
 
 
+def find_idx_of_first_nonzero(stm, start, max_, left):
+    if left:
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py]: This function is unused. I removed it.

File: tutorials/nlp/Punctuation_and_Capitalization.ipynb (hunk 1 above)
Comment (PeganovAnton) [reply tutorials/nlp/Punctuation_and_Capitalization.ipynb]: I removed the example

File: examples/nlp/token_classification/punctuate_capitalize.py (hunk 2 above)
Comment (PeganovAnton) [reply examples/nlp/token_classification/punctuate_capitalize.py]: I written to the script description that this values are optimal for IWSLT 2019 test dataset.

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 6]
```diff
@@ -0,0 +1,739 @@
...
+        list(itertools.chain(*all_subtokens_mask)),
+        list(itertools.chain(*all_quantities_of_preceding_words)),
+        list(itertools.chain(*all_query_ids)),
+        list(itertools.chain(*all_is_first)),
+        list(itertools.chain(*all_is_last)),
+    )
+
+
+class BertPunctuationCapitalizationInferLongDataset(Dataset):
+    """
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:315]: @PeganovAnton @AlexGrinch let's move this to `https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py`. This inference script is way too long and most of the functions could be hidden

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 7]
```diff
@@ -0,0 +1,739 @@
...
+        stm = _remove_margins(stm, margin, keep_left=first, keep_right=last)
+        for b_probs, logits in [(b_punct_probs, pl), (b_capit_probs, cl)]:
+            p = torch.nn.functional.softmax(
+                _remove_margins(logits, margin, keep_left=first, keep_right=last)[stm], dim=-1,
+            )
+            b_probs.append(p.detach().cpu().numpy())
+    return b_punct_probs, b_capit_probs, new_start_word_ids
+
+
+def add_punctuation_capitalization(
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:602]: With this, the model will have two implementations of the same function, I think it's better to keep as it was originally for this PR: extend the functionally of the function and keep only a single version inside the model class

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+        if capit_label != cfg.dataset.pad_label:
+            word = word.capitalize()
+        query_with_punct_and_capit += word
+        if punct_label != cfg.dataset.pad_label:
+            query_with_punct_and_capit += punct_label
+        query_with_punct_and_capit += ' '
+    return query_with_punct_and_capit[:-1]
+
+
+def _transform_logit_to_prob_and_remove_margins_and_extract_word_probs(
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:554]: move to the model class or create a utils file inside https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/nlp/models/token_classification

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 9]
```diff
@@ -0,0 +1,739 @@
...
+            f"Not enough accumulated probabilities. Number_of_probs_to_move={number_of_probs_to_move} "
+            f"acc_prob.shape={acc_prob.shape}"
+        )
+    if number_of_probs_to_move > 0:
+        pred = pred + list(np.argmax(acc_prob[:number_of_probs_to_move], axis=-1))
+    acc_prob = acc_prob[number_of_probs_to_move:]
+    return pred, acc_prob
+
+
+def _update_accumulated_probabilities(acc_prob: np.ndarray, update: np.ndarray) -> np.ndarray:
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:499]: move to the model class or create a utils file inside https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/nlp/models/token_classification

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 10]
```diff
@@ -0,0 +1,739 @@
...
+        update: numpy array of shape ``[A + N, L]``
+    Returns:
+        numpy array of shape ``[A + N, L]``
+    """
+    acc_prob *= update[: acc_prob.shape[0]]
+    acc_prob = np.concatenate([acc_prob, update[acc_prob.shape[0] :]], axis=0)
+    return acc_prob
+
+
+def _remove_margins(tensor, margin_size, keep_left, keep_right):
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:512]: move to the model class or create a utils file inside https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/nlp/models/token_classification

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 11]
```diff
@@ -0,0 +1,739 @@
...
+    if not keep_left:
+        tensor = tensor[margin_size + 1 :]  # remove left margin and CLS token
+    if not keep_right:
+        tensor = tensor[: tensor.shape[0] - margin_size - 1]  # remove right margin and SEP token
+    return tensor
+
+
+def apply_punct_capit_predictions(cfg: DictConfig, query: str, punct_preds: List[int], capit_preds: List[int]) -> str:
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:520]: move to the model class or create a utils file inside https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/nlp/models/token_classification

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 12]
```diff
@@ -0,0 +1,739 @@
...
+def load_manifest(manifest: Path):
+    result = []
+    with manifest.open() as f:
+        for i, line in enumerate(f):
+            data = json.loads(line)
+            result.append(data)
+    return result
+
+
+def _get_subtokens_and_subtokens_mask(query: str, tokenizer: TokenizerSpec) -> Tuple[List[str], List[int]]:
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:173]: move to the model class or create a utils file inside https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/nlp/models/token_classification

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 13]
```diff
@@ -0,0 +1,739 @@
...
+    subtokens_mask = []
+    for j, word in enumerate(words):
+        word_tokens = tokenizer.text_to_tokens(word)
+        subtokens.extend(word_tokens)
+        subtokens_mask.append(1)
+        subtokens_mask.extend([0] * (len(word_tokens) - 1))
+    return subtokens, subtokens_mask
+
+
+def _check_max_seq_length_and_margin_and_step(max_seq_length: int, margin: int, step: int):
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:196]: move to the model class or create a utils file inside https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/nlp/models/token_classification

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 14]
```diff
@@ -0,0 +1,739 @@
...
+    if step <= 0:
+        raise ValueError(f"Parameter `step` has to be positive whereas step={step}")
+    if step > max_seq_length - 2 - 2 * margin:
+        logging.warning(
+            f"Parameter step={step} is too big. It will be reduced to `min(max_seq_length, <maximum query length> + 2) "
+            f"- 2 - 2 * margin`."
+        )
+
+
+def get_features_infer(
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:227]: move to the `dataset` class or create a utils file inside https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/nlp/models/token_classification

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 15]
```diff
@@ -0,0 +1,739 @@
...
+            np.array(self.all_input_mask[idx], dtype=np.float32),
+            np.array(self.all_subtokens_mask[idx]),
+            self.all_quantities_of_preceding_words[idx],
+            self.all_query_ids[idx],
+            self.all_is_first[idx],
+            self.all_is_last[idx],
+        )
+
+
+def setup_infer_long_dataloader(
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:428]: move to the model class

File: examples/nlp/token_classification/punctuate_capitalize_infer.py [hunk 16]
```diff
@@ -0,0 +1,739 @@
...
+        collate_fn=dataset.collate_fn,
+        batch_size=batch_size,
+        shuffle=False,
+        num_workers=model._cfg.dataset.num_workers,
+        pin_memory=model._cfg.dataset.pin_memory,
+        drop_last=False,
+    )
+
+
+def _move_acc_probs_to_token_preds(
```
Comment (ekmb) [inline examples/nlp/token_classification/punctuate_capitalize_infer.py:473]: move to the model class or create a utils file inside https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/nlp/models/token_classification

**Review by ekmb — CHANGES_REQUESTED** [review]

**Review by ekmb — COMMENTED** [review]
@PeganovAnton please add a small ci/unit test as it was in the initial PR version.

File: examples/nlp/token_classification/punctuate_capitalize_infer.py (hunk 13 above)
Comment (PeganovAnton) [reply examples/nlp/token_classification/punctuate_capitalize_infer.py]: Moved to dataset file

File: examples/nlp/token_classification/punctuate_capitalize_infer.py (hunk 12 above)
Comment (PeganovAnton) [reply examples/nlp/token_classification/punctuate_capitalize_infer.py]: moved to dataset file

File: examples/nlp/token_classification/punctuate_capitalize_infer.py (hunk 14 above)
Comment (PeganovAnton) [reply examples/nlp/token_classification/punctuate_capitalize_infer.py]: moved to dataset file

**Review by ekmb — DISMISSED** [review]

**Review by ekmb — APPROVED** [review]

**Pull request merged**